### PR TITLE
Implement Single Tap mod for Taiko

### PR DESCRIPTION
**Pull request by OpenSauce04**
This PR implements the Single Tap mod in Taiko
This is implemented such that Don and Kat hits are treated as independent instances of how the Single Tap mod works for osu!standard, with each action type only being allowed to use the first key that activates it.
Strong hits are allowed to be hit with both the left and right actions of its respective type.
I recorded a demonstration here: https://streamable.com/2mqmna

I also wanted to implement the Alternate mod, however what "alternating" in Taiko actually _is_ seems to be quite hotly debated, so I decided to leave that for a future PR

Apologies if I missed anything that needs to be added when implementing a mod, I'm a bit unfamiliar with the mod system

### CODE REVIEW

**Comment by OpenSauce04** [comment]
Also, I believe a test scene needs to be added for this, however I have no idea how the test scene for the Standard single tap mod works

**Comment by mk56-spn** [comment]
> Also, I believe a test scene needs to be added for this, however I have no idea how the test scene for the Standard single tap mod works

I'll get around to making a test scene this afternoon at some point if you don't figure it out

File: osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            }
+
+            nonGameplayPeriods = new PeriodTracker(periods);
+
+            gameplayClock = drawableRuleset.FrameStableClock;
+        }
+
+        public void Update(Playfield playfield)
+        {
+            if (nonGameplayPeriods.IsInAny(gameplayClock.CurrentTime))
```
Comment (mk56-spn) [inline osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:72]: would invert this if, less nesting usually makes for an easier read, and since its the only code within the update, there's no reason no to IMO

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+using osu.Framework.Localisation;
+
+namespace osu.Game.Rulesets.Taiko.Mods
+{
+    public class TaikoModSingleTap : TaikoInputBlockingMod
+    {
+        public override string Name => @"Single Tap";
+        public override string Acronym => @"SG";
+        public override LocalisableString Description => @"You must only use one key!";
+
```
Comment (mk56-spn) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs:13]: this is speculative and id like a second opinion since the added code isn't much regardless but maybe there should be a base mod class for single tap between rulesets to avoid this duplicated code

File: osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            {
+                case TaikoAction.LeftCentre:
+                case TaikoAction.RightCentre:
+                case TaikoAction.LeftRim:
+                case TaikoAction.RightRim:
+                    break;
+
+                // Any action which is not left or right button should be ignored.
+                default:
+                    return true;
```
Comment (mk56-spn) [inline osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:99]: Not all that familiar with taiko, but is the number of cases that arent these cases larger or smaller? , if there's less of them then the default should likely be these, and have the exceptions listed.

**Review by mk56-spn — CHANGES_REQUESTED** [review]
Very light code quality pass, will add anything related to functionality after i do the tests

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs (hunk 2 above)
Comment (OpenSauce04) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs:13]: I thought of this too, but I'm unsure of how effective this would be given that almost all areas of the code had *some* kind of change to it that was Taiko-specific

File: osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs (hunk 1 above)
Comment (mk56-spn) [reply osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:72]: its a nitpick so i wouldn't worry about fixing it there as well since it brings unrelated files in.

File: osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs (hunk 3 above)
Comment (OpenSauce04) [reply osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:99]: Sorry could you rephrase that? I'm not sure what you mean

Comment (mk56-spn) [reply osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:99]: > Sorry could you rephrase that? I'm not sure what you mean

of the actions not listed there is there any? since i cant see any other ones in the `taikoaction` enum, so im not sure this safeguard is neccesary

Comment (OpenSauce04) [reply osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:99]: It seems that this is a leftover from Stable to make sure that you can still use smoke
I'll remove it

File: osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs (hunk 1 above)
Comment (OpenSauce04) [reply osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:70]: > would invert this if

Sorry, but how would I do this?
I'm not sure exactly what changes you are describing

**Comment by OpenSauce04** [comment]
Additional note, should the description for Single Tap be changed on Taiko? You aren't using one key, you are using two

**Comment by mk56-spn** [comment]
> Additional note, should the description for Single Tap be changed on Taiko? You aren't using one key, you are using two

probably, something like "only allows one key per input type".

**Comment by OpenSauce04** [comment]
> probably, something like "only allows one key per input type". 
How about "One key for dons, one key for kats"

**Comment by mk56-spn** [comment]
> > probably, something like "only allows one key per input type".
> 
> How about "One key for dons, one key for kats" Fits in well with the existing mod descriptions

wasnt sure if there was precedent for the usage of those names but seeing as there is then yeah

**Comment by mk56-spn** [comment]
aside from the tests which i will provide, the mod compatibilities need to be checkedhttps://github.com/ppy/osu/pull/22746/checks?check_run_id=11629614869

**Comment by OpenSauce04** [comment]
> aside from the tests which i will provide, the mod compatibilities need to be checked https://github.com/ppy/osu/pull/22746/checks?check_run_id=11629614869

Yeah I forgot to do that and noticed that in the tests while I was away from my PC, I'll get to that in a couple of hours

File: osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs (hunk 1 above)
Comment (bdach) [reply osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:70]:
```csharp
public void Method()
{
    if (condition)
    {
        // entire actual body of the method
    }
}
```
vs
```csharp
public void Method()
{
    if (!condition) return;
    // entire body of the method
}
```

Comment (OpenSauce04) [reply osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:70]: Ah, I see what you mean now. Thanks!

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs (hunk 2 above)
Comment (bdach) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs]: I think the potential harm outweights the benefit. Extracting a base class for three strings is not much gain, and having them separate prevents the possibility of overcomplicated class hierarchies popping up.

If there is a class to commonise, then it can be done after the pattern of exactly what is needed is clear, not before.

File: osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+using osu.Game.Rulesets.Taiko.Objects;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Rulesets.UI;
+using osu.Game.Screens.Play;
+using osu.Game.Utils;
+using osu.Game.Rulesets.Taiko.UI;
+
+namespace osu.Game.Rulesets.Taiko.Mods
+{
+    public abstract partial class TaikoInputBlockingMod : Mod, IApplicableToDrawableRuleset<TaikoHitObject>, IUpdatableByPlayfield
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:22]: I'm skeptical of this structure. For the initial implementation I would not do `TaikoInputBlockingMod` at all, but implement the single tap mod in a single file. No inheritance (except from `Mod`, obviously).

The rationale is that I don't know what the alternate mod will look like, and I have no way of knowing if the structure put down here is going to nicely extend to mod. In fact I have major doubts it will; `LastAccepted{Don,Kat}Action` likely have no place in this base class once people start with their kkdd/kddk/ddkk alternating madness.

Comment (OpenSauce04) [reply osu.Game.Rulesets.Taiko/Mods/TaikoInputBlockingMod.cs:22]: Sounds reasonable, I'll get around to it

**Comment by mk56-spn** [comment]
> @mk56-spn I believe the only main thing left is implementing the test; If you are still offering to do it, I would hugely appreciate it 🙂

Yeah will do. Did most of it yesterday but then saw you had to do a refactor so I let it sit

**Comment by mk56-spn** [comment]
Will try to have the tests done in the next couple of hours

**Comment by mk56-spn** [comment]
https://user-images.githubusercontent.com/74463310/221988494-196c5393-d019-4ab1-b89f-2c9d2c28ffc2.mov

`TestScene` is mostly done but there seems to be an issue, you seem to be blocking inputs only if they come directly from gameplay, where as the osu ruleset singletap blocks inputs from replays as well, as can be seen here in the test which uses a fake replay to simulate taps all the inputs go through regardless of whether they were single tapped or alternated

**Comment by mk56-spn** [comment]
Relevant code

```// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
// See the LICENCE file in the repository root for full licence text.

using System.Collections.Generic;
using NUnit.Framework;
using osu.Game.Beatmaps;
using osu.Game.Rulesets.Objects;
using osu.Game.Rulesets.Replays;
using osu.Game.Rulesets.Taiko.Mods;
using osu.Game.Rulesets.Taiko.Objects;
using osu.Game.Rulesets.Taiko.Replays;

namespace osu.Game.Rulesets.Taiko.Tests.Mods
{
    public partial class TestSceneTaikoModSingleTap : TaikoModTestScene
    {
        [Test]
        public void TestInputAlternate() => CreateModTest(new ModTestData
        {
            Mod = new TaikoModSingleTap(),
            Autoplay = false,
            Beatmap = new Beatmap
            {
                HitObjects = new List<HitObject>
                {
                    new Hit
                    {
                        StartTime = 100,
                        Type = HitType.Rim
                    },
                    new Hit
                    {
                        StartTime = 300,
                        Type = HitType.Rim
                    },
                    new Hit
                    {
                        StartTime = 500,
                        Type = HitType.Rim
                    },
                    new Hit
                    {
                        StartTime = 700,
                        Type = HitType.Rim
                    },
                },
            },
            ReplayFrames = new List<ReplayFrame>
            {
                new TaikoReplayFrame(100, TaikoAction.RightRim),
                new TaikoReplayFrame(120),
                new TaikoReplayFrame(300, TaikoAction.LeftRim),
                new TaikoReplayFrame(320),
                new TaikoReplayFrame(500, TaikoAction.RightRim),
                new TaikoReplayFrame(520),
                new TaikoReplayFrame(700, TaikoAction.LeftRim),
                new TaikoReplayFrame(720),
            },
            PassCondition = () => Player.ScoreProcessor.Combo.Value == 4
        });

        [Test]
        public void TestInputSameKey() => CreateModTest(new ModTestData
        {
            Mod = new TaikoModSingleTap(),
            Autoplay = false,
            Beatmap = new Beatmap
            {
                HitObjects = new List<HitObject>
                {
                    new Hit
                    {
                        StartTime = 100,
                        Type = HitType.Rim
                    },
                    new Hit
                    {
                        StartTime = 300,
                        Type = HitType.Rim
                    },
                    new Hit
                    {
                        StartTime = 500,
                        Type = HitType.Rim
                    },
                    new Hit
                    {
                        StartTime = 700,
                        Type = HitType.Rim
                    },
                },
            },
            ReplayFrames = new List<ReplayFrame>
            {
                new TaikoReplayFrame(100, TaikoAction.RightRim),
                new TaikoReplayFrame(120),
                new TaikoReplayFrame(300, TaikoAction.RightRim),
                new TaikoReplayFrame(320),
                new TaikoReplayFrame(500, TaikoAction.RightRim),
                new TaikoReplayFrame(520),
                new TaikoReplayFrame(700, TaikoAction.RightRim),
                new TaikoReplayFrame(720),
            },
            PassCondition = () => Player.ScoreProcessor.Combo.Value == 4
        });
    }
}

**Comment by OpenSauce04** [comment]
Thanks!!
I'll take a look at the replay thing

**Comment by OpenSauce04** [comment]
Sorry, I've been busy for a couple of days, I'll finish this PR off tomorrow morning

**Comment by OpenSauce04** [comment]
Alright should be ready for review now

**Comment by mk56-spn** [comment]
Worth noting I removed two tests that were present in the osu mod single tap whilst making the taiko tests. Adding them back should be as simple as copy pasting them if necessary and tweaking them to match the existing two ( ie using taiko replay inputs, and pointing to the taiko mod )

**Comment by OpenSauce04** [comment]
Alright, I added the two missing tests, as well as an extra test for strong hit objects

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        public override double ScoreMultiplier => 1.0;
+        public override Type[] IncompatibleMods => new[] { typeof(ModAutoplay), typeof(ModRelax), typeof(TaikoModCinema) };
+        public override ModType Type => ModType.Conversion;
+
+        private DrawableTaikoRuleset ruleset = null!;
+
+        private TaikoPlayfield playfield { get; set; } = null!;
+
+        protected TaikoAction? LastAcceptedDonAction { get; private set; }
+        protected TaikoAction? LastAcceptedKatAction { get; private set; }
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs:48]: - Please make `private` for now
- Can these be named `LastAccepted{Centre,Rim}Action`? I would rather avoid having to mentally parse "is centre don or kat again?" every time this code is read.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+        /// <summary>
+        /// A tracker for periods where alternate should not be forced (i.e. non-gameplay periods).
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs:51]: Comment refers to "alternate" wrongly

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+        public void Update(Playfield playfield)
+        {
+            if (!nonGameplayPeriods.IsInAny(gameplayClock.CurrentTime)) return;
+
+            if (LastAcceptedDonAction != null)
+                LastAcceptedDonAction = null;
+
+            if (LastAcceptedKatAction != null)
+                LastAcceptedKatAction = null;
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs:91]: Nit: Null checks serve no purpose and can be removed (as in, unconditionally assign `null`)

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+        protected bool CheckValidNewAction(TaikoAction action)
+        {
+            if (action == TaikoAction.LeftCentre || action == TaikoAction.RightCentre)
+                return LastAcceptedDonAction == null || LastAcceptedDonAction == action;
+            if (action == TaikoAction.LeftRim || action == TaikoAction.RightRim)
+                return LastAcceptedKatAction == null || LastAcceptedKatAction == action;
+
+            return true;
+        }
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs:37]: Please inline this into `checkCorrectAction()` for now. The fact that this is split and has the same checks duplicated that that method does is making this some very arduous and confusing reading.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by OpenSauce04** [comment]
I've made the requested changes

**Review by bdach — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]
